--- pristine.snapshot.lean	2026-06-30 13:06:48
+++ ../p3_koala_bear.lean	2026-06-30 13:14:03
@@ -2,6 +2,7 @@
 -- Experimental lean backend for Hax
 -- The Hax prelude library can be found in hax/proof-libs/lean
 import Hax
+import p3_koala_bear.dependencies
 import Std.Tactic.Do
 import Std.Do.Triple
 import Std.Tactic.Do.Syntax
@@ -17,11 +18,6 @@
 structure KoalaBearParameters where
   -- no fields
 
---  The prime field `2^31 - 2^24 + 1`, a.k.a. the Koala Bear field.
-abbrev KoalaBear :
-  Type :=
-  (p3_monty_31.monty_31.MontyField31 KoalaBearParameters)
-
 @[instance] opaque Impl_11.AssociatedTypes :
   core_models.clone.Clone.AssociatedTypes KoalaBearParameters :=
   by constructor <;> exact Inhabited.default
@@ -99,6 +95,11 @@
   MONTY_BITS := (32 : u32)
   MONTY_MU := (2164260865 : u32)
 
+--  The prime field `2^31 - 2^24 + 1`, a.k.a. the Koala Bear field.
+abbrev KoalaBear :
+  Type :=
+  (p3_monty_31.monty_31.MontyField31 KoalaBearParameters)
+
 @[reducible] instance Impl_1.AssociatedTypes :
   p3_challenger.duplex_challenger.UniformSamplingField.AssociatedTypes
   KoalaBearParameters
@@ -148,7 +149,7 @@
           (pure (rust_primitives.hax.Tuple2.mk a k)) :
           RustM (rust_primitives.hax.Tuple2 (RustArray u64 64) usize))));
     (pure a))
-    (by rfl)
+    (by sorry)
 
 @[reducible] instance Impl_2.AssociatedTypes :
   p3_monty_31.data_traits.PackedMontyParameters.AssociatedTypes
@@ -176,7 +177,7 @@
   where
   MONTY_GEN := RustM.of_isOk
     (do (p3_monty_31.monty_31.Impl.new KoalaBearParameters (3 : u32)))
-    (by rfl)
+    (by sorry)
 
 @[reducible] instance Impl_5.AssociatedTypes :
   p3_monty_31.data_traits.RelativelyPrimePower.AssociatedTypes
@@ -234,7 +235,7 @@
                               (542991299 : u32),
                               (1760025929 : u32),
                               (1791270792 : u32)])))))
-    (by rfl)
+    (by sorry)
   ROOTS_8 := RustM.of_isOk
     (do
     (rust_primitives.unsize
@@ -243,7 +244,7 @@
                               (1748172362 : u32),
                               (2113994754 : u32),
                               (391001680 : u32)])))))
-    (by rfl)
+    (by sorry)
   INV_ROOTS_8 := RustM.of_isOk
     (do
     (rust_primitives.unsize
@@ -252,7 +253,7 @@
                               (1739704753 : u32),
                               (16711679 : u32),
                               (382534071 : u32)])))))
-    (by rfl)
+    (by sorry)
   ROOTS_16 := RustM.of_isOk
     (do
     (rust_primitives.unsize
@@ -265,7 +266,7 @@
                               (982097957 : u32),
                               (391001680 : u32),
                               (668978722 : u32)])))))
-    (by rfl)
+    (by sorry)
   INV_ROOTS_16 := RustM.of_isOk
     (do
     (rust_primitives.unsize
@@ -278,7 +279,7 @@
                               (1464983071 : u32),
                               (382534071 : u32),
                               (1982081381 : u32)])))))
-    (by rfl)
+    (by sorry)
 
 @[reducible] instance Impl_7.AssociatedTypes :
   p3_monty_31.data_traits.BinomialExtensionData.AssociatedTypes
@@ -296,29 +297,17 @@
   where
   W := RustM.of_isOk
     (do (p3_monty_31.monty_31.Impl.new KoalaBearParameters (3 : u32)))
-    (by rfl)
-  mul_w :=
-    fun
-      (A : Type)
-      [trait_constr_mul_w_associated_type_i0 :
-        p3_field.field.Algebra.AssociatedTypes
-        A
-        (p3_monty_31.monty_31.MontyField31 KoalaBearParameters)]
-      [trait_constr_mul_w_i0 : p3_field.field.Algebra
-        A
-        (p3_monty_31.monty_31.MontyField31 KoalaBearParameters)
-        ] (a : A) => do
-    (core_models.ops.arith.Add.add
-      A
-      A (← (p3_field.field.PrimeCharacteristicRing.double A a)) a)
+    (by sorry)
+  mul_w := sorry
+
   DTH_ROOT := RustM.of_isOk
     (do (p3_monty_31.monty_31.Impl.new KoalaBearParameters (2113994754 : u32)))
-    (by rfl)
+    (by sorry)
   EXT_GENERATOR := RustM.of_isOk
     (do
     (p3_monty_31.monty_31.Impl.new_array KoalaBearParameters ((4 : usize))
       (RustArray.ofVec #v[(2 : u32), (1 : u32), (0 : u32), (0 : u32)])))
-    (by rfl)
+    (by sorry)
   EXT_TWO_ADICITY := (26 : usize)
   TWO_ADIC_EXTENSION_GENERATORS := RustM.of_isOk
     (do
@@ -334,7 +323,7 @@
                                                   (0 : u32),
                                                   (0 : u32),
                                                   (777715144 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 @[reducible] instance Impl_8.AssociatedTypes :
   p3_monty_31.data_traits.BinomialExtensionData.AssociatedTypes
@@ -352,10 +341,10 @@
   where
   W := RustM.of_isOk
     (do (p3_monty_31.monty_31.Impl.new KoalaBearParameters (3 : u32)))
-    (by rfl)
+    (by sorry)
   DTH_ROOT := RustM.of_isOk
     (do (p3_monty_31.monty_31.Impl.new KoalaBearParameters (1748172362 : u32)))
-    (by rfl)
+    (by sorry)
   EXT_GENERATOR := RustM.of_isOk
     (do
     (p3_monty_31.monty_31.Impl.new_array KoalaBearParameters ((8 : usize))
@@ -367,7 +356,7 @@
                             (0 : u32),
                             (0 : u32),
                             (0 : u32)])))
-    (by rfl)
+    (by sorry)
   EXT_TWO_ADICITY := (27 : usize)
   TWO_ADIC_EXTENSION_GENERATORS := RustM.of_isOk
     (do
@@ -399,21 +388,8 @@
                                                   (0 : u32),
                                                   (0 : u32),
                                                   (14348907 : u32)])])))
-    (by rfl)
-  mul_w :=
-    fun
-      (A : Type)
-      [trait_constr_mul_w_associated_type_i0 :
-        p3_field.field.Algebra.AssociatedTypes
-        A
-        (p3_monty_31.monty_31.MontyField31 KoalaBearParameters)]
-      [trait_constr_mul_w_i0 : p3_field.field.Algebra
-        A
-        (p3_monty_31.monty_31.MontyField31 KoalaBearParameters)
-        ] (a : A) => do
-    (core_models.ops.arith.Add.add
-      A
-      A (← (p3_field.field.PrimeCharacteristicRing.double A a)) a)
+    (by sorry)
+  mul_w := sorry
 
 @[reducible] instance Impl_9.AssociatedTypes :
   p3_monty_31.data_traits.TrinomialQuinticData.AssociatedTypes
@@ -434,7 +410,7 @@
                             (0 : u32),
                             (0 : u32),
                             (0 : u32)])))
-    (by rfl)
+    (by sorry)
   EXT_TWO_ADICITY := (24 : usize)
   FROBENIUS_COEFFS := RustM.of_isOk
     (do
@@ -462,10 +438,10 @@
                                                   (2008416047 : u32),
                                                   (143367062 : u32),
                                                   (1027410849 : u32)])])))
-    (by rfl)
+    (by sorry)
   TWO_ADIC_EXTENSION_GENERATORS := RustM.of_isOk
     (do (pure (RustArray.ofVec #v[])))
-    (by rfl)
+    (by sorry)
 
 end p3_koala_bear.koala_bear
 
@@ -507,7 +483,7 @@
                             (3 : i64),
                             (4 : i64),
                             (9 : i64)])))
-    (by rfl)
+    (by sorry)
   MATRIX_CIRC_MDS_12_COL := RustM.of_isOk
     (do
     (p3_mds.util.first_row_to_first_col ((12 : usize)) i64
@@ -523,7 +499,7 @@
                             (9 : i64),
                             (4 : i64),
                             (10 : i64)])))
-    (by rfl)
+    (by sorry)
   MATRIX_CIRC_MDS_16_COL := RustM.of_isOk
     (do
     (p3_mds.util.first_row_to_first_col ((16 : usize)) i64
@@ -543,7 +519,7 @@
                             (22 : i64),
                             (13 : i64),
                             (3 : i64)])))
-    (by rfl)
+    (by sorry)
   MATRIX_CIRC_MDS_24_COL := RustM.of_isOk
     (do
     (p3_mds.util.first_row_to_first_col ((24 : usize)) i64
@@ -571,7 +547,7 @@
                             (1551339770 : i64),
                             (400627958 : i64),
                             (142123135 : i64)])))
-    (by rfl)
+    (by sorry)
   MATRIX_CIRC_MDS_32_COL := RustM.of_isOk
     (do
     (p3_mds.util.first_row_to_first_col ((32 : usize)) i64
@@ -607,7 +583,7 @@
                             (1415143558 : i64),
                             (1216108043 : i64),
                             (1755622571 : i64)])))
-    (by rfl)
+    (by sorry)
   MATRIX_CIRC_MDS_64_COL := RustM.of_isOk
     (do
     (p3_mds.util.first_row_to_first_col ((64 : usize)) i64
@@ -675,7 +651,7 @@
                             (488082023 : i64),
                             (1488175179 : i64),
                             (1855310673 : i64)])))
-    (by rfl)
+    (by sorry)
 
 abbrev MdsMatrixKoalaBear :
   Type :=
@@ -689,10 +665,7 @@
 --  External (full round) layer for KoalaBear Poseidon1.
 abbrev Poseidon1ExternalLayerKoalaBear (WIDTH : usize) :
   Type :=
-  (p3_monty_31.no_packing.poseidon1.Poseidon1ExternalLayerMonty31
-    p3_koala_bear.koala_bear.KoalaBearParameters
-    p3_koala_bear.mds.MDSKoalaBearData
-    (WIDTH))
+  sorry
 
 --  S-box degree for KoalaBear Poseidon1.
 -- 
@@ -744,29 +717,14 @@
 --  Internal (partial round) layer for KoalaBear Poseidon1.
 abbrev Poseidon1InternalLayerKoalaBear (WIDTH : usize) :
   Type :=
-  (p3_monty_31.no_packing.poseidon1.Poseidon1InternalLayerMonty31
-    p3_koala_bear.koala_bear.KoalaBearParameters
-    (WIDTH)
-    KoalaBearPoseidonParameters)
+  sorry
 
 --  The Poseidon1 permutation for KoalaBear.
 -- 
 --  Acts on arrays of the form `[KoalaBear; WIDTH]` or `[KoalaBear::Packing; WIDTH]`.
 abbrev Poseidon1KoalaBear (WIDTH : usize) :
   Type :=
-  (p3_poseidon1.Poseidon1
-    (p3_monty_31.monty_31.MontyField31
-      p3_koala_bear.koala_bear.KoalaBearParameters)
-    (p3_monty_31.no_packing.poseidon1.Poseidon1ExternalLayerMonty31
-      p3_koala_bear.koala_bear.KoalaBearParameters
-      p3_koala_bear.mds.MDSKoalaBearData
-      (WIDTH))
-    (p3_monty_31.no_packing.poseidon1.Poseidon1InternalLayerMonty31
-      p3_koala_bear.koala_bear.KoalaBearParameters
-      (WIDTH)
-      KoalaBearPoseidonParameters)
-    (WIDTH)
-    ((3 : u64)))
+  sorry
 
 --  Generic Poseidon1 linear layers for KoalaBear.
 -- 
@@ -1352,7 +1310,7 @@
                                                   (616885102 : u32),
                                                   (334644983 : u32),
                                                   (132306927 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-24 Poseidon1 on KoalaBear.
 -- 
@@ -2120,7 +2078,7 @@
                                                   (307342664 : u32),
                                                   (1962630877 : u32),
                                                   (570045382 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 --  Create a default width-16 Poseidon1 permutation for KoalaBear.
 @[spec]
@@ -2213,9 +2171,7 @@
 
 abbrev Poseidon2ExternalLayerKoalaBear (WIDTH : usize) :
   Type :=
-  (p3_monty_31.no_packing.poseidon2.Poseidon2ExternalLayerMonty31
-    p3_koala_bear.koala_bear.KoalaBearParameters
-    (WIDTH))
+  sorry
 
 --  Number of full rounds per half for KoalaBear Poseidon2 (`RF / 2`).
 -- 
@@ -2339,7 +2295,7 @@
                                                   (64412292 : u32),
                                                   (1936878279 : u32),
                                                   (1980661727 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-16 Poseidon2 on KoalaBear.
 -- 
@@ -2427,7 +2383,7 @@
                                                   (1585100155 : u32),
                                                   (867006515 : u32),
                                                   (879151050 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-16 Poseidon2 on KoalaBear.
 -- 
@@ -2468,7 +2424,7 @@
                             (185193011 : u32),
                             (452207447 : u32),
                             (510054082 : u32)])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-24 Poseidon2 on KoalaBear.
 -- 
@@ -2588,7 +2544,7 @@
                                                   (1608418116 : u32),
                                                   (1083269213 : u32),
                                                   (691256798 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-24 Poseidon2 on KoalaBear.
 -- 
@@ -2708,7 +2664,7 @@
                                                   (1757946479 : u32),
                                                   (1551204074 : u32),
                                                   (681266718 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-24 Poseidon2 on KoalaBear.
 -- 
@@ -2752,7 +2708,7 @@
                             (213359415 : u32),
                             (603124968 : u32),
                             (1038411577 : u32)])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-32 Poseidon2 on KoalaBear.
 -- 
@@ -2904,7 +2860,7 @@
                                                   (1208969182 : u32),
                                                   (1478187151 : u32),
                                                   (440934505 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-32 Poseidon2 on KoalaBear.
 -- 
@@ -3056,7 +3012,7 @@
                                                   (1718116838 : u32),
                                                   (1790111045 : u32),
                                                   (1501131602 : u32)])])))
-    (by rfl)
+    (by sorry)
 
 --  Round constants for width-32 Poseidon2 on KoalaBear.
 -- 
@@ -3108,7 +3064,7 @@
                             (2121419385 : u32),
                             (667840239 : u32),
                             (1995134368 : u32)])))
-    (by rfl)
+    (by sorry)
 
 --  Contains data needed to define the internal layers of the Poseidon2 permutation.
 structure KoalaBearInternalLayerParameters where
@@ -3116,10 +3072,7 @@
 
 abbrev Poseidon2InternalLayerKoalaBear (WIDTH : usize) :
   Type :=
-  (p3_monty_31.no_packing.poseidon2.Poseidon2InternalLayerMonty31
-    p3_koala_bear.koala_bear.KoalaBearParameters
-    (WIDTH)
-    KoalaBearInternalLayerParameters)
+  sorry
 
 --  An implementation of the Poseidon2 hash function specialised to run on the current architecture.
 -- 
@@ -3127,18 +3080,7 @@
 --  wherever possible, input arrays should of the form `[KoalaBear::Packing; WIDTH]`.
 abbrev Poseidon2KoalaBear (WIDTH : usize) :
   Type :=
-  (p3_poseidon2.Poseidon2
-    (p3_monty_31.monty_31.MontyField31
-      p3_koala_bear.koala_bear.KoalaBearParameters)
-    (p3_monty_31.no_packing.poseidon2.Poseidon2ExternalLayerMonty31
-      p3_koala_bear.koala_bear.KoalaBearParameters
-      (WIDTH))
-    (p3_monty_31.no_packing.poseidon2.Poseidon2InternalLayerMonty31
-      p3_koala_bear.koala_bear.KoalaBearParameters
-      (WIDTH)
-      KoalaBearInternalLayerParameters)
-    (WIDTH)
-    ((3 : u64)))
+  sorry
 
 --  An implementation of the matrix multiplications in the internal and external layers of Poseidon2.
 -- 
@@ -3189,178 +3131,8 @@
   p3_koala_bear.koala_bear.KoalaBearParameters
   ((16 : usize))
   where
-  internal_layer_mat_mul :=
-    fun
-      (R : Type)
-      [trait_constr_internal_layer_mat_mul_associated_type_i0 :
-        p3_field.field.PrimeCharacteristicRing.AssociatedTypes
-        R]
-      [trait_constr_internal_layer_mat_mul_i0 :
-        p3_field.field.PrimeCharacteristicRing
-        R
-        ] (state : (RustArray R 16)) (sum : R) => do
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (1 : usize)
-        (← (core_models.ops.arith.AddAssign.add_assign
-          R
-          R (← state[(1 : usize)]_?) (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (2 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R (← state[(2 : usize)]_?)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (3 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.halve
-            R (← state[(3 : usize)]_?)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (4 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (core_models.ops.arith.Add.add
-            R
-            R
-            (← (p3_field.dup.Dup.dup R sum))
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(4 : usize)]_?)))))
-          (← (p3_field.dup.Dup.dup R (← state[(4 : usize)]_?))))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (5 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(5 : usize)]_?))))))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (6 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.halve
-            R (← state[(6 : usize)]_?))))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (7 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (core_models.ops.arith.Add.add
-            R
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(7 : usize)]_?)))
-            (← (p3_field.dup.Dup.dup R (← state[(7 : usize)]_?))))))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (8 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(8 : usize)]_?))))))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (9 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(9 : usize)]_?) (8 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (10 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(10 : usize)]_?) (3 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (11 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(11 : usize)]_?) (24 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (12 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(12 : usize)]_?) (8 : u64))))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (13 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(13 : usize)]_?) (3 : u64))))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (14 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(14 : usize)]_?) (4 : u64))))));
-    let state : (RustArray R 16) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (15 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          sum
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(15 : usize)]_?) (24 : u64))))));
-    (pure state)
+  internal_layer_mat_mul := sorry
 
---  Create a default width-16 Poseidon2 permutation for KoalaBear.
 @[spec]
 def default_koalabear_poseidon2_16 (_ : rust_primitives.hax.Tuple0) :
     RustM
@@ -3425,258 +3197,8 @@
   p3_koala_bear.koala_bear.KoalaBearParameters
   ((24 : usize))
   where
-  internal_layer_mat_mul :=
-    fun
-      (R : Type)
-      [trait_constr_internal_layer_mat_mul_associated_type_i0 :
-        p3_field.field.PrimeCharacteristicRing.AssociatedTypes
-        R]
-      [trait_constr_internal_layer_mat_mul_i0 :
-        p3_field.field.PrimeCharacteristicRing
-        R
-        ] (state : (RustArray R 24)) (sum : R) => do
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (1 : usize)
-        (← (core_models.ops.arith.AddAssign.add_assign
-          R
-          R (← state[(1 : usize)]_?) (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (2 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R (← state[(2 : usize)]_?)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (3 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.halve
-            R (← state[(3 : usize)]_?)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (4 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (core_models.ops.arith.Add.add
-            R
-            R
-            (← (p3_field.dup.Dup.dup R sum))
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(4 : usize)]_?)))))
-          (← (p3_field.dup.Dup.dup R (← state[(4 : usize)]_?))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (5 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(5 : usize)]_?))))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (6 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.halve
-            R (← state[(6 : usize)]_?))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (7 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (core_models.ops.arith.Add.add
-            R
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(7 : usize)]_?)))
-            (← (p3_field.dup.Dup.dup R (← state[(7 : usize)]_?))))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (8 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(8 : usize)]_?))))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (9 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(9 : usize)]_?) (8 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (10 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(10 : usize)]_?) (2 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (11 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(11 : usize)]_?) (3 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (12 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(12 : usize)]_?) (4 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (13 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(13 : usize)]_?) (5 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (14 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(14 : usize)]_?) (6 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (15 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(15 : usize)]_?) (24 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (16 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(16 : usize)]_?) (8 : u64))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (17 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(17 : usize)]_?) (3 : u64))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (18 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(18 : usize)]_?) (4 : u64))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (19 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(19 : usize)]_?) (5 : u64))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (20 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(20 : usize)]_?) (6 : u64))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (21 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(21 : usize)]_?) (7 : u64))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (22 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(22 : usize)]_?) (9 : u64))))));
-    let state : (RustArray R 24) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (23 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          sum
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(23 : usize)]_?) (24 : u64))))));
-    (pure state)
+  internal_layer_mat_mul := sorry
 
---  Create a default width-24 Poseidon2 permutation for KoalaBear.
 @[spec]
 def default_koalabear_poseidon2_24 (_ : rust_primitives.hax.Tuple0) :
     RustM
@@ -3769,338 +3291,8 @@
   p3_koala_bear.koala_bear.KoalaBearParameters
   ((32 : usize))
   where
-  internal_layer_mat_mul :=
-    fun
-      (R : Type)
-      [trait_constr_internal_layer_mat_mul_associated_type_i0 :
-        p3_field.field.PrimeCharacteristicRing.AssociatedTypes
-        R]
-      [trait_constr_internal_layer_mat_mul_i0 :
-        p3_field.field.PrimeCharacteristicRing
-        R
-        ] (state : (RustArray R 32)) (sum : R) => do
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (1 : usize)
-        (← (core_models.ops.arith.AddAssign.add_assign
-          R
-          R (← state[(1 : usize)]_?) (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (2 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R (← state[(2 : usize)]_?)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (3 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.halve
-            R (← state[(3 : usize)]_?)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (4 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (core_models.ops.arith.Add.add
-            R
-            R
-            (← (p3_field.dup.Dup.dup R sum))
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(4 : usize)]_?)))))
-          (← (p3_field.dup.Dup.dup R (← state[(4 : usize)]_?))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (5 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(5 : usize)]_?))))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (6 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.halve
-            R (← state[(6 : usize)]_?))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (7 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (core_models.ops.arith.Add.add
-            R
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(7 : usize)]_?)))
-            (← (p3_field.dup.Dup.dup R (← state[(7 : usize)]_?))))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (8 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.double
-            R
-            (← (p3_field.field.PrimeCharacteristicRing.double
-              R (← state[(8 : usize)]_?))))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (9 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(9 : usize)]_?) (8 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (10 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(10 : usize)]_?) (2 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (11 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(11 : usize)]_?) (3 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (12 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(12 : usize)]_?) (4 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (13 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(13 : usize)]_?) (5 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (14 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(14 : usize)]_?) (6 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (15 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(15 : usize)]_?) (10 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (16 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(16 : usize)]_?) (12 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (17 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(17 : usize)]_?) (14 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (18 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(18 : usize)]_?) (16 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (19 : usize)
-        (← (core_models.ops.arith.Add.add
-          R
-          R
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(19 : usize)]_?) (24 : u64)))
-          (← (p3_field.dup.Dup.dup R sum)))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (20 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(20 : usize)]_?) (8 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (21 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(21 : usize)]_?) (3 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (22 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(22 : usize)]_?) (4 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (23 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(23 : usize)]_?) (5 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (24 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(24 : usize)]_?) (6 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (25 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(25 : usize)]_?) (7 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (26 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(26 : usize)]_?) (9 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (27 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(27 : usize)]_?) (10 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (28 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(28 : usize)]_?) (12 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (29 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(29 : usize)]_?) (14 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (30 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          (← (p3_field.dup.Dup.dup R sum))
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(30 : usize)]_?) (16 : u64))))));
-    let state : (RustArray R 32) ←
-      (rust_primitives.hax.monomorphized_update_at.update_at_usize
-        state
-        (31 : usize)
-        (← (core_models.ops.arith.Sub.sub
-          R
-          R
-          sum
-          (← (p3_field.field.PrimeCharacteristicRing.div_2exp_u64
-            R (← state[(31 : usize)]_?) (24 : u64))))));
-    (pure state)
+  internal_layer_mat_mul := sorry
 
---  Create a default width-32 Poseidon2 permutation for KoalaBear.
 @[spec]
 def default_koalabear_poseidon2_32 (_ : rust_primitives.hax.Tuple0) :
     RustM
